java/arcs/core/allocator/Arc.kt

```diff
@@ -151,8 +153,14 @@ class Arc internal constructor(
             deferred.complete(this@Arc)
         }
 
-        fetchCurrentStates()
-        recomputeArcState()
+        val shouldFire = mutex.withLock {
```

Comment:
I know testing threading issues is tricky, but can we add some basic tests aiming to gain confidence around in particular risk of deadlocks?

Reply:
Only three things hold this mutex lock

1) on callback from ArcHost, it updates state, clones a list, and then releases the lock. No change of deadlock. 

2) the initial registration and subsequent onArcStateChanged lock the mutex while they retrieve the current states and update the combined arcState. These ArcHost calls have a timeout enforced of 5000ms beore releasing the lock. 

I guess I could make a mock ArcHost to try and trip up a race, by making the lookupState() calls take a long time but making the callbacks spam quickly, but I feel pretty confidence there's no deadlock here given the aforementioned 5s timeout which should guarantee the mutexes are released.

I could run this with --runs_per_test=10000 if that would increase your confidence?

Reply:
I support landing this but I think we should add a test as next step, something like what you describe with many threads sounds worthwhile. Perhaps worth getting input from @ianull as well with ideas, maybe we could add something to his existing test apps to look at this case.

Reply:
I support getting this PR in as well.
The lock is quite straightforward here to protect state as well as listener callbacks.
As for the stability test currently we have, sadly I have to say: 
it's a random crash monkey test on `Storage` stack and `Arc/Allocator` layers are not touched at all :-(
Perhaps there is a need to add a task to extend the stability test to cover Arc/Allocator for Q3.